persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -180,6 +180,20 @@ createPostgresqlPoolModifiedWithVersion
 createPostgresqlPoolModifiedWithVersion getVer modConn ci =
   createSqlPool $ open' modConn getVer ci
 
+createPostgresqlPoolWithConf
+    :: (MonadUnliftIO m, MonadLogger m)
+    => (PG.Connection -> IO (Maybe Double)) -- ^ Action to perform to get the server version.
+    -> (PG.Connection -> IO ()) -- ^ Action to perform after connection is created.
```

Comment:
These two functions are kind of annoying, naming wise. Something like `createPostgresqlPoolModifiedWithVersionWithConf` would be gross (`createPostgresqlPoolModifiedWithVersion` is pretty bad on its own tbh). I'm also somewhat worried that another callback will eventually be added here (Perhaps a callback for when a connection is destroyed, or some other Postgres specific thing?)

Possibly a separate data type is called for to hold these functions (`PostgresConfCallbacks` isn't quite right, but something like that), with a function to create a default data structure?

Reply:
`Hooks` is another term that gets used for this concept a bit - 

```haskell
data PostgresConfHooks = PostgresConfHooks
  { getServerVersion :: IO (Maybe Double)
  , afterCreate :: IO ()
  }

createPostgresqlPoolWithHooks :: (PG.Connection -> PostgresConfHooks) -> ...
```

Reply:
So, I was investigating this today and I was thinking this could be an opportunity to clean up the server version API. I have these complaints:

1. Double is a weird way to represent a version. It doesn't work for 3 digit versions (x.y.z) (though Postgres moved to just 2 digit versions recently which helps). I think something like `NonEmpty Integer` makes more sense, e.g. `[10, 2]`

2. The `Maybe` behavior feels weird to me. If you return `Nothing`, Persistent assumes the lowest possible version, not trying to do things like upsert. This isn't documented afaict, and would subtly degrade behavior if someone happened to return `Nothing`.

As long as we're adding a new API, I would propose something like:

```
getServerVersion :: IO (NonEmpty Integer)
```

In the case of Redshift users who can't get the server version (assuming that's still a thing), they can return a hard-coded value that matches whatever Postgres capabilities Redshift has, e.g. `pure (10 :| [2])`.

If there was something better suited than `NonEmpty Integer` I'd be fine with that too.

Reply:
Implementation would be something like:

```
getServerVersion2 :: PG.Connection -> IO (NonEmpty Word)
getServerVersion2 conn = do
  [PG.Only version] <- PG.query_ conn "show server_version";
  case AT.parseOnly parseVersion (T.pack version) of
    Left err -> throwIO $ PostgresServerVersionError $ "Failed to parse Postgres version. Got: " <> version <> ". Error: " <> err
    Right versionComponents -> case NEL.nonEmpty versionComponents of
      Nothing -> throwIO $ PostgresServerVersionError $ "Empty Postgres version string. Got: " <> version
      Just neVersion -> pure neVersion

  where
    -- Partially copied from the `versions` package
    parseVersion = AT.decimal `AT.sepBy` AT.char '.' <* AT.endOfInput
```

Reply:
Sounds great to me